### ISSUE TITLE
Optional Callable behaves oddly

### ISSUE THREAD

**Issue by BradXiao** [microsoft/pyright#3512]
**Describe the bug**
A variable is defined as optional callable cannot be analyzed properly.

**To Reproduce**
Refer to the code below.

**Expected behavior**
In the provided code, such scenario should work properly.

**Screenshots or Code**
##### Consistent case
![圖片](https://user-images.githubusercontent.com/20597547/170403160-b74bc50a-50db-4445-9a92-b0767eab4a05.png)

##### Inconsistent case
![圖片](https://user-images.githubusercontent.com/20597547/170403171-ef274baa-0a7d-4f3d-b5d3-61dbb3c96ec5.png)

```python
from typing import Callable, List, Optional

class TextRegion():
    x:int
    y:int
    text:str

def get_something(reg:List[TextRegion],condition:Optional[Callable]=None):
    new_reg=None
    for i in range(len(reg)-1):
        for j in range(i+1,len(reg)):
            if reg[i].x==0.7:
                if reg[i].x > reg[j].x and\
                     (condition is None or (condition(reg[i]) and condition(reg[j]))  ):
                    new_reg=reg[j]
                    
                    break
            elif reg[j].x>=0.8:
                if (condition is None or (condition(reg[i]) and condition(reg[j]))):
                    new_reg=reg[j]

                    break

    return new_reg
```

**Additional context**
Pylance extension = v2022.5.4-pre.1
Python extension = v2022.6.2
VS code = 1.67.2
OS = Windows 10

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem. I'll investigate further.

Here's a stripped-down sample that repros the problem.

```python
from typing import Callable

def func(call: Callable | None):
    while True:
        while True:
            if call is None:
                break
            if call([]):
                break
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.250, which I just published. It will also be included in the next release of pylance.
